Various updates

nickryall opened the pull request:
Apologies for the large pull request. I had been making these changes while working on a project and now am unsure how to submit the individual commits as requests. If you would prefer them as individual requests let me know and I'll manually break them up.

This pull-request contains the following:

**Disable options by model attribute**

When using a collection I needed a way to disable options based on a boolean attribute set in the model. 

**Return 'valid' if the the value is empty and the field is *not* set to 'required'.**

If the select field is optional then the form should still submit if for example the default 'unselectedText' option is selected.

**Set the name attribute on the select element with the passed in 'name' option.**

If the select element in the markup does not have it's name attribute set. Use the name passed in through options. 

**Clear method**

Resetting the form ( e.g.  via a type="reset" button ) does not clear out the value set on the field. So I created a 'clear' method to do this manually.  I suspect I may be missing something here though and there is probably an easier way to achieve this. 

**Note:** I've done the same for ampersand-input-view and also a clearFields convenience method for ampersand-form-view which will call reset() on all child fields. If this is something that makes it in I'll submit the pull requests for these also.

nickryall commented:
Any feedback on this? Would it be better to close this ticket and submit an individual PR for each item?

wraithgar commented:
I think this is probably all worth doing at once. Can you sync this w/ master again? There was some code added that affects the select.name that we may now need to think more closely on when we set it (this.name <=> this.select.name)

nickryall commented:
This is now synced with master.

kamilogorek commented:
@nickryall code looks fine, however we need tests coverage and docs update for `disabled` functionality you've added. Would you mind writing them?

nickryall commented:
@kamilogorek Sure thing. I'll get onto it.

nickryall commented:
1. Added documentation and tests for disabled functionality.  

2. Removed default of 'disabled' for the disabledAttribute option. This means the user will have to explicitly set an attribute here. This will prevent the possible case where a model already has a 'disabled' attribute but disabling the resulting select option is not desired.

3. Updated tests to take into account change #2 above (Return 'valid' if the the value is empty and the field is not set to 'required').

All tests pass using testem locally and in the pre-commit check. 

The Travis CI build is failing however:

Error:
Zuul tried to run tests in saucelabs, however no saucelabs credentials were provided.
See the zuul wiki (https://github.com/defunctzombie/zuul/wiki/Cloud-testing) for info on how to setup cloud testing.

kamilogorek commented:
@nickryall I guess you forgot to add documentation updates to git commit :)

kamilogorek commented:
Regarding Travis build, cc: @bear

bear commented:
i've updated the travis-ci environment config @kamilogorek @nickryall

kamilogorek commented:
Thanks @bear!

nickryall commented:
@kamilogorek  Yep I hadn't saved the readme in my editor on the first commt ;) But it's there in second  'Readme' commit  above your comment.

kamilogorek commented on ampersand-select-view.js:
```diff
@@ -196,31 +202,39 @@ SelectView.prototype.setValue = function (value) {
     this.value = value;
-    this.validate();
+
+    if(!skipValidation) {
```
missing whitespace

kamilogorek commented on ampersand-select-view.js:
```diff
@@ -196,31 +202,39 @@ SelectView.prototype.setValue = function (value) {
-                return this.options.indexOf(this.value) > -1;
-            } else {
-                return !!this.findModelForId(this.value);
+    if(!this.value && !this.required) {
```
here as well

kamilogorek commented on ampersand-select-view.js:
```diff
@@ -196,31 +202,39 @@ SelectView.prototype.setValue = function (value) {
+        this.valid = true;
+    } else {
+        this.valid = this.options.some(function (element) {
+
```
redundant newline

kamilogorek commented on ampersand-select-view.js:
```diff
@@ -133,7 +135,7 @@ SelectView.prototype.renderOptions = function () {
     this.options.forEach(function (option) {
         this.select.appendChild(
-            createOption(this.getOptionValue(option), this.getOptionText(option))
+            createOption(this.getOptionValue(option), this.getOptionText(option), this.getOptionDisabled(option))
```
missing semicolon

nickryall replied:
@kamilogorek The createOption call is an argument for the appendChild method so I don't think there should be a semicolon in this instance?

kamilogorek commented on ampersand-select-view.js:
```diff
@@ -274,12 +302,16 @@ SelectView.prototype.getOptionText = function (option) {
     //Set to empty-string if undefined or null, but not if 0, false, etc
     if (value === null || value === undefined) { value = ''; }
 
+    if(disabled) {
```
missing whitespace

kamilogorek commented on ampersand-select-view.js:
```diff
@@ -243,6 +257,20 @@ SelectView.prototype.getOptionValue = function (option) {
 };
 
+SelectView.prototype.getOptionDisabled = function (option) {
+    if (Array.isArray(option)) return;
```
If you'll take a look at `getOptionValue` and `getOptionText`, you'll see that it should return `option[1]`;

nickryall replied:
@kamilogorek The disabled functionality does not currently work for non model items. So if the type is array I think it should simply return here. For items which are an array ( when sending pairs as the options), getOptionValue returns option[0] and getOptionText returns option[1]. option[2] would be undefined. 

We could however extend the functionality for sending arrays to allow a 3rd value which would toggle disabled on the option. e.g.

    options: [ ['a', 'Option A', true], ['b', 'Option B', false], ['c', 'Option C', true] ]

kamilogorek commented on ampersand-select-view.js:
```diff
@@ -243,6 +257,20 @@ SelectView.prototype.getOptionValue = function (option) {
+    if (Array.isArray(option)) return;
+
+    if (this.options.isCollection) {
+        if (this.disabledAttribute && option[this.disabledAttribute]) {
```
This can be reduced to

```js
if (this.disabledAttribute) return option[this.disabledAttribute];
```

As it'll return exactly same logical thing.

kamilogorek commented:
@nickryall +1 from me after applying those small patches :)

nickryall commented:
1. Cleaned up the whitespace and single line if statements. 
2. Enabled the disabled functionality when using an array for option creation. 
3. I also realised there were no tests for using arrays for option creation so I added those.

kamilogorek commented:
Magnificent :+1:

wraithgar commented:
I'm +1 on this, minor version?

kamilogorek commented:
Agree.

wraithgar commented:
v2.4.0
